### PR TITLE
Dev ontoraster and geospatial analysis

### PR DESCRIPTION
**Pull request by jc2341**

### CODE REVIEW

File: Deploy/stacks/UK-building-retrofit/StackDeployment/OBDA mappings/OBDA_BuildingEPC.txt
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/OBDA mappings/OBDA_BuildingEPC.txt:1]: Thanks! This issue has been addressed in commit [6d878bc0cdd35541c842ff1ef195c7e5639b9f19]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/SPARQL queries/EPC_Avg_Query_For_All_Areas.txt.txt
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/SPARQL queries/EPC_Avg_Query_For_All_Areas.txt.txt:1]: Thank you for the reminder. This issue is addressed in commit [ffcb33006f28f2ae53754a62a6e8324b44330d7a]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-data-uploader/data/uk_geomgraphic_hierarchy/tabular/readme.txt
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-data-uploader/data/uk_geomgraphic_hierarchy/tabular/readme.txt:1]: This typo has been fixed, and the associated 'datasetDirectory' settings in Stack-data-uploader/config has also been examined. Please refer to commit [3d20e6c8e86859abb219b373eea0d1d044a13522] for more details

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/SERVICE_CONFIG_FILES_HAVE_MOVED.md
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/SERVICE_CONFIG_FILES_HAVE_MOVED.md:1]: Thank you for your careful examination. This file was removed in commit [3d20e6c8e86859abb219b373eea0d1d044a13522]

File: Deploy/stacks/UK-building-retrofit/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# UK Building Retrofit
+
+This repository provides a set of instructions for creating an interoperable environment to integrate UK-wide building energy performance, energy consumption, fuel poverty, and public health records. The use case is to quantify building retrofit urgency and determine optimal retrofit pathways across multiple geographic scales.
```
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/README.md]: Thank you for the clarification. I have updated the term 'repository' to 'project folder' or 'folder' where appropriate to ensure clarity. Please check commit [612db81f48585f4caa826a062f2caefd620932c6]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Settings of remote Linux server/README.md
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Settings of remote Linux server/README.md:1]: This file was removed, please check commit [352a9cea8e4845e866d18b1cd8d86995e6049c42]

File: Deploy/stacks/dynamic/stack-data-uploader/inputs/config/TOID_UPRN_MatchingTable.json
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/dynamic/stack-data-uploader/inputs/config/TOID_UPRN_MatchingTable.json:1]: Very sorry for this. The project files are removed. Please check commit [8b8b179341492867f6a823e44a3f760ebf642a6c]

File: Deploy/stacks/agent/Linked-Identfier/UPRN_Matching.py
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/agent/Linked-Identfier/UPRN_Matching.py:1]: Sorry for the oversight. This content is stale and has been removed. Please see the changes in commit [041dec9304cdc12063b82c601f34e2766e4bf9ee].

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore:1]: Thank you for pointing this out. The issue was caused by my improper use of .gitignore. The updated version of the in-use file has already been included in commit [1b46188ec8b714fcf57c0abef25605e6861b5dbc].

File: Deploy/stacks/agent/Linked-Identfier/UPRN_Matching.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import psycopg2
+import requests
+from tqdm import tqdm
+
+# Function to get UPRN from API by Given OSGB
+def get_uprn(osgb):
+    api_key = 'aXLg2jRUgjmtRobP73q8gWSp3dtku05f'
```
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/agent/Linked-Identfier/UPRN_Matching.py]: Thank you for your careful observation. This was an outdated operation from last August. The matching and querying of UPRN and TOID have since been unified within ONTOP. As a result, this part has already been removed in commit [041dec9304cdc12063b82c601f34e2766e4bf9ee]

File: Deploy/stacks/agent/GDB-Uploader/Cambridge_Ely/.gitignore
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/agent/GDB-Uploader/Cambridge_Ely/.gitignore:1]: Thank you for pointing this out. The code in this folder (e.g. GDP_uploader ) was previously used for batch uploading geodatabase (.gdb) files for 3D buildings. However, this functionality has become redundant since stack-data-uploader incorporated GDP since December 2023. As a result, all related files have been removed in commit [041dec9304cdc12063b82c601f34e2766e4bf9ee].

File: Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore:1]: Sorry for the oversight. This folder has been cleaned in commit [8b8b179341492867f6a823e44a3f760ebf642a6c]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/feature-info-agent.json
[the remark this answers is not in the document]
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/feature-info-agent.json:1]: Yes, this JSON file in the stack folder is unnecessary and was left behind during my first practice. Since that, I have already set FIA as one of the services in the stack, but it was excluded from the repository earlier due to an improper .gitignore configuration. Now, the updated code with FIA as a service is available by commit [1b46188ec8b714fcf57c0abef25605e6861b5dbc], and I have verified that it works after deleting the redundant JSON file.

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-cesium.json
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-cesium.json:1]: Is this Cesium visualisation being used at all?

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-mapbox.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+    "ServiceSpec": {
+        "Name": "vis-mapbox",
+        "TaskTemplate": {
+            "ContainerSpec": {
+                "Image": "ghcr.io/cambridge-cares/dtvf-base-image:latest",
```
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-mapbox.json:6]: This is hopelessly outdated. Is this deliberate? I thought you had upgraded everything to the latest TWA vis framework?

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/feature-info-agent.json
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/feature-info-agent.json:1]: I guess I didn't manage to explain it clearly enough in my previous comment. The FIA is now (and has been for quite a while) a _BUILT-IN_ service of the stack. This means that no `feature-info-agent.json` file is required in any `config/services` folder. Please delete it.

In order to spin up the FIA in a stack, all you need to do is add its name in the relevant stack manager input config `json` file.

File: Deploy/stacks/UK-building-retrofit/StackDeployment/README.md [hunk 4]
```diff
@@ -117,21 +116,32 @@ This query retrieves the average energy efficiency ratings for all buildings in
...
+[SSH tunnel]: https://code.visualstudio.com/docs/remote/tunnels
+[Feature Info Agent]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/FeatureInfoAgent
 
 <!-- Stack references -->
 [common stack scripts]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Deploy/stacks/dynamic/common-scripts
 [Stack data uploader]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Deploy/stacks/dynamic/stack-data-uploader
 [Stack manager]: https://github.com/cambridge-cares/TheWorldAvatar/blob/main/Deploy/stacks/dynamic/stack-manager/README.md
+[fia_quires]: ./Stack-manager/inputs/data/fia-queries
+[services]: ./Stack-manager/inputs/config/services
+[feature-info-agent.json]: ./Stack-manager/inputs/config/services/feature-info-agent.json
```
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/README.md:145]: This does not and should not exist. Please see my other comment regarding the FIA.

File: Deploy/stacks/UK-building-retrofit/StackDeployment/README.md [hunk 5]
```diff
@@ -117,21 +116,32 @@ This query retrieves the average energy efficiency ratings for all buildings in
...
+[The World Avatar]: https://github.com/cambridge-cares/TheWorldAvatar.git
+[git clone]: https://git-scm.com/docs/git-clone
+[SSH tunnel]: https://code.visualstudio.com/docs/remote/tunnels
+[Feature Info Agent]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Agents/FeatureInfoAgent
 
 <!-- Stack references -->
 [common stack scripts]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Deploy/stacks/dynamic/common-scripts
 [Stack data uploader]: https://github.com/cambridge-cares/TheWorldAvatar/tree/main/Deploy/stacks/dynamic/stack-data-uploader
 [Stack manager]: https://github.com/cambridge-cares/TheWorldAvatar/blob/main/Deploy/stacks/dynamic/stack-manager/README.md
+[fia_quires]: ./Stack-manager/inputs/data/fia-queries
```
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/README.md:143]: Please fix the typo.

File: Deploy/stacks/UK-building-retrofit/StackDeployment/README.md [hunk 6]
```diff
@@ -117,21 +116,32 @@ This query retrieves the average energy efficiency ratings for all buildings in
 
 Further query templates can be found in the Deploy/stacks/UK-building-retrofit/StackDeployment/SPARQL queries directory, covering a range of metrics such as energy consumption, fuel poverty rates, and public health data.
 
+## 5. Feature Info Agent
+The [Feature Info Agent] serves as an access point for the visualisation, enabling queries about data for individual features (e.g., specific geographical locations). In this case, the Feature Info Agent retrieves energy performance and structural descriptions for buildings by querying the Ontop endpoint using the building's IRI.  
+
+The queries and config files can be found in [fia_quires]. Navigate to [services], open [feature-info-agent.json] and under the `Mounts` section, modify `Source` based on where the queries and config files are located at.
```
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/README.md:122]: The bit "Navigate to..." is obsolete. Please remove it. Please see my other comments regarding the FIA.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-cesium.json
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-cesium.json:1]: No, we usually choose Mapbox. I included Cesium earlier because the visualization tutorial covered both. However, for practical purposes, I have removed Cesium in commit [90c1f8a22ae01d944b04da511a367baced4c7333]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-mapbox.json (hunk 3 above)
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/vis-mapbox.json]: Apologies, I copied the wrong visualization backup. I’ve now uploaded the version using TWA VF 4.6.0 in commit [4b17b6014b2f81e5f29eeb93a7c99e4cb941202f] and commit [e62212b3beebc69c3143e987516269629e6b716b]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/feature-info-agent.json
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/feature-info-agent.json:1]: Thanks for point it out. I have corrected the service folder and stack json file. This time FIA will built-in and use query stored in default folder, no hardcode problems. Please see commit [4b17b6014b2f81e5f29eeb93a7c99e4cb941202f]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/README.md (hunk 5 above)
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/README.md]: Sorry for the mistake, this was fixed in commit [4eb60b4acc6e48d628673b8595c1cd90a6c490f9]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/README.md (hunk 4 above)
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/README.md]: Thanks, this is addressed in commit [4eb60b4acc6e48d628673b8595c1cd90a6c490f9]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/README.md (hunk 6 above)
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/README.md]: Thank you for the reminder, this is addressed in commit [4eb60b4acc6e48d628673b8595c1cd90a6c490f9] and commit [ 840cd640264d6d66e2c1f301372c8b6640399878]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/visualisation.json
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/visualisation.json:1]: Just like the FIA, the visualisation is now a _built-in_ service of the stack, and therefore no `json` config for this service is required. Please remove this file.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/visualisation.json
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/visualisation.json:1]: Thanks for pointing it out! This was addressed in commit [4dc832e027da30c648668bcb93526851cec0277e].

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore
Comment (sm453) [inline Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore:1]: Why did you re-add this file? Please remove this. You don't need this folder at all. There should not be any customised services required.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore
Comment (sm453) [inline Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore:1]: Please revert this change. As commented before, no changes should be made to core stack infrastructure files on this branch.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore
Comment (sm453) [inline Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore:1]: Please _REVERT_ this change. That means undo it. Simply deleting the file is not appropriate. We do need it.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore
Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore:1]: Sorry I just saw this in some project and mistakenly thought it was neseccary here. Noe this was removed in commit [14e158788f09d3bbcddf04f3eefa1c6e04cb7368]

Comment (jc2341) [reply Deploy/stacks/UK-building-retrofit/StackDeployment/Stack-manager/inputs/config/services/.gitignore:1]: Sorry, I just saw some other project has this, I mistakenly thought it was necessary here. This was addressed in commit [14e158788f09d3bbcddf04f3eefa1c6e04cb7368]

File: Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore
Comment (jc2341) [reply Deploy/stacks/dynamic/stack-data-uploader/inputs/data/.gitignore:1]: Thanks for the guidance, the core stack (stack manager and stack uploader) were replaced with the current version in main, details are in commit [de154f41d8f39d27b351b82f770b27410033ba9c]

**Review by sm453 — APPROVED** [review]
